File: python/examples/crew-ai-cdp-chatbot/chatbot.py
import asyncio
import json
import os
import sys
import time

from crewai import Agent, Crew, Task
from dotenv import load_dotenv

from coinbase_agentkit import (
    AgentKit,
    AgentKitConfig,
    CdpEvmWalletProvider,
    CdpEvmWalletProviderConfig,
    X402Config,
    cdp_api_action_provider,
    cdp_evm_wallet_action_provider,
    erc20_action_provider,
    pyth_action_provider,
    wallet_action_provider,
    weth_action_provider,
    x402_action_provider,
)
from coinbase_agentkit_crewai import get_crewai_tools


def initialize_agent(config: CdpEvmWalletProviderConfig) -> tuple[Agent, CdpEvmWalletProvider]:
    """Initialize a CrewAI agent with CDP AgentKit tools."""
    wallet_provider = CdpEvmWalletProvider(
        CdpEvmWalletProviderConfig(
            api_key_id=config.api_key_id,
            api_key_secret=config.api_key_secret,
            wallet_secret=config.wallet_secret,
            network_id=config.network_id,
            address=config.address,
            idempotency_key=config.idempotency_key,
            rpc_url=config.rpc_url,
        )
    )

    agentkit = AgentKit(
        AgentKitConfig(
            wallet_provider=wallet_provider,
            action_providers=[
                cdp_api_action_provider(),
                cdp_evm_wallet_action_provider(),
                erc20_action_provider(),
                pyth_action_provider(),
                wallet_action_provider(),
                weth_action_provider(),
                x402_action_provider(
                    X402Config(
                        registered_services=["https://www.x402.org/protected"]
                        if config.network_id == "base-sepolia"
                        else [],
                        allow_dynamic_service_registration=False,
                        max_payment_usdc=1.0,
                        registered_facilitators={
                            "my-custom-facilitator": "https://my-custom-facilitator.example.com",
                        },
                    )
                ),
            ],
        )
    )

    tools = get_crewai_tools(agentkit)

    agent = Agent(
        role="CDP Agent",
        goal="Help users interact onchain with Coinbase Developer Platform AgentKit tools.",
        backstory=(
            "You are a concise and careful onchain assistant. Before executing your first "
            "action, check wallet details so you know which network you are using. If the "
            "wallet needs funds on base-sepolia, use the faucet. If a request cannot be "
            "completed with the available tools, say so and point the user to "
            "docs.cdp.coinbase.com."
        ),
        llm=os.getenv("OPENAI_MODEL", "gpt-4o-mini"),
        tools=tools,
        verbose=True,
        allow_delegation=False,
    )

    return agent, wallet_provider


def setup() -> Agent:
    """Set up the CrewAI agent with persistent wallet storage."""
    network_id = os.getenv("NETWORK_ID", "base-sepolia")
    wallet_file = f"wallet_data_{network_id.replace('-', '_')}.txt"

    wallet_data = {}
    if os.path.exists(wallet_file):
        try:
            with open(wallet_file) as f:
                wallet_data = json.load(f)
                print(f"Loading existing wallet from {wallet_file}")
        except json.JSONDecodeError:
            print(f"Warning: Invalid wallet data for {network_id}")
            wallet_data = {}

    wallet_address = wallet_data.get("address") or os.getenv("ADDRESS") or None

    config = CdpEvmWalletProviderConfig(
        api_key_id=os.getenv("CDP_API_KEY_ID"),
        api_key_secret=os.getenv("CDP_API_KEY_SECRET"),
        wallet_secret=os.getenv("CDP_WALLET_SECRET"),
        network_id=network_id,
        address=wallet_address,
        idempotency_key=(os.getenv("IDEMPOTENCY_KEY") if not wallet_address else None),
        rpc_url=os.getenv("RPC_URL"),
    )

    agent, wallet_provider = initialize_agent(config)

    new_wallet_data = {
        "address": wallet_provider.get_address(),
        "network_id": network_id,
        "created_at": time.strftime("%Y-%m-%d %H:%M:%S")
        if not wallet_data
        else wallet_data.get("created_at"),
    }

    with open(wallet_file, "w") as f:
        json.dump(new_wallet_data, f, indent=2)
        print(f"Wallet data saved to {wallet_file}")

    return agent


def run_prompt(agent: Agent, prompt: str) -> str:
    """Run a single user prompt through a one-task CrewAI crew."""
    task = Task(
        description=f"Handle this user request with the available AgentKit tools:\n\n{prompt}",
        expected_output="A concise response with any relevant transaction hashes, links, or wallet details.",
        agent=agent,
    )
    crew = Crew(agents=[agent], tasks=[task], verbose=True)
    return str(crew.kickoff())


async def run_autonomous_mode(agent: Agent, interval: int = 10) -> None:
    """Run the agent autonomously with a delay between tasks."""
    print("Starting autonomous mode...")
    while True:
        try:
            output = run_prompt(
                agent,
                "Choose one useful onchain action that highlights your abilities.",
            )
            print(output)
            print("-------------------")
            await asyncio.sleep(interval)
        except KeyboardInterrupt:
            print("Goodbye Agent!")
            sys.exit(0)


async def run_chat_mode(agent: Agent) -> None:
    """Run the agent interactively based on user input."""
    print("Starting chat mode... Type 'exit' to end.")
    while True:
        try:
            user_input = input("\nPrompt: ")
            if user_input.lower() == "exit":
                break

            print(run_prompt(agent, user_input))
            print("-------------------")
        except KeyboardInterrupt:
            print("Goodbye Agent!")
            sys.exit(0)


def choose_mode() -> str:
    """Choose whether to run in autonomous or chat mode based on user input."""
    while True:
        print("\nAvailable modes:")
        print("1. chat    - Interactive chat mode")
        print("2. auto    - Autonomous action mode")

        choice = input("\nChoose a mode (enter number or name): ").lower().strip()
        if choice in ["1", "chat"]:
            return "chat"
        if choice in ["2", "auto"]:
            return "auto"
        print("Invalid choice. Please try again.")


async def main() -> None:
    """Start the chatbot agent."""
    load_dotenv()

    agent = setup()

    print("\nWelcome to the CDP AgentKit CrewAI Chatbot!")
    print("Type 'exit' to quit the chat.\n")

    mode = choose_mode()
    if mode == "chat":
        await run_chat_mode(agent=agent)
    elif mode == "auto":
        await run_autonomous_mode(agent=agent)


if __name__ == "__main__":
    print("Starting Agent...")
    asyncio.run(main())

File: typescript/examples/vercel-ai-sdk-waap-chatbot/chatbot.ts
import {
  AgentKit,
  WaapWalletProvider,
  EvmWalletProvider,
  customActionProvider,
  erc20ActionProvider,
  pythActionProvider,
  walletActionProvider,
  wethActionProvider,
} from "@coinbase/agentkit";
import { getVercelAITools } from "@coinbase/agentkit-vercel-ai-sdk";
import { openai } from "@ai-sdk/openai";
import { streamText, ToolSet, stepCountIs } from "ai";
import * as dotenv from "dotenv";
import * as fs from "fs";
import * as readline from "readline";
import { z } from "zod";

dotenv.config();

const WALLET_DATA_FILE = "wallet_data.txt";

const SignMessageSchema = z.object({
  message: z.string().describe("The message to sign."),
});

const SignTypedDataSchema = z.object({
  typedData: z.record(z.string(), z.any()).optional().describe("Full EIP-712 typed data object."),
  typedDataJson: z.string().optional().describe("Stringified JSON for EIP-712 typed data."),
  domain: z.record(z.string(), z.any()).optional().describe("EIP-712 domain object."),
  types: z
    .record(z.string(), z.array(z.object({ name: z.string(), type: z.string() })))
    .optional()
    .describe("EIP-712 type definitions."),
  primaryType: z.string().optional().describe("Primary type name for typed data."),
  message: z.record(z.string(), z.any()).optional().describe("EIP-712 message payload."),
});

const ReadContractSchema = z.object({
  contractAddress: z.string().describe("The contract address to query."),
  abi: z.array(z.record(z.string(), z.any())).describe("Contract ABI array."),
  functionName: z.string().describe("Name of the view/pure function to call."),
  args: z.array(z.any()).optional().describe("Function arguments."),
});

const waapAdvancedActions = customActionProvider<EvmWalletProvider>([
  {
    name: "sign_message",
    description:
      "Sign an arbitrary message with the connected WaaP wallet using personal_sign semantics.",
    schema: SignMessageSchema,
    invoke: async (walletProvider, args) => {
      const signature = await walletProvider.signMessage(args.message);
      return `Message signature: ${signature}`;
    },
  },
  {
    name: "sign_typed_data",
    description:
      "Sign EIP-712 typed data with the connected WaaP wallet. Input must include domain, types, primaryType, and message.",
    schema: SignTypedDataSchema,
    invoke: async (walletProvider, args) => {
      let typedDataPayload: Record<string, unknown> | undefined = args.typedData;

      if (!typedDataPayload && args.typedDataJson) {
        typedDataPayload = JSON.parse(args.typedDataJson) as Record<string, unknown>;
      }

      if (!typedDataPayload && args.domain && args.types && args.primaryType && args.message) {
        typedDataPayload = {
          domain: args.domain,
          types: args.types,
          primaryType: args.primaryType,
          message: args.message,
        };
      }

      if (!typedDataPayload) {
        throw new Error(
          "Missing typed data. Provide typedData, typedDataJson, or domain/types/primaryType/message.",
        );
      }

      const signature = await walletProvider.signTypedData(typedDataPayload);
      return `Typed data signature: ${signature}`;
    },
  },
  {
    name: "read_contract",
    description:
      "Read a pure/view function from a smart contract using contract address, ABI, function name, and args.",
    schema: ReadContractSchema,
    invoke: async (walletProvider, args) => {
      const result = await walletProvider.readContract({
        address: args.contractAddress as `0x${string}`,
        abi: args.abi,
        functionName: args.functionName,
        args: args.args ?? [],
      });
      return `Contract read result: ${JSON.stringify(result, (_key, value) =>
        typeof value === "bigint" ? value.toString() : value,
      )}`;
    },
  },
]);

/**
 * Validates required environment variables.
 */
function validateEnvironment(): void {
  const missingVars: string[] = [];

  const requiredVars = ["OPENAI_API_KEY", "WAAP_EMAIL", "WAAP_PASSWORD"];
  requiredVars.forEach(varName => {
    if (!process.env[varName]) {
      missingVars.push(varName);
    }
  });

  if (missingVars.length > 0) {
    console.error("Error: Required environment variables are not set");
    missingVars.forEach(varName => {
      console.error(`${varName}=your_${varName.toLowerCase()}_here`);
    });
    process.exit(1);
  }

  if (!process.env.WAAP_CHAIN_ID) {
    console.warn("Warning: WAAP_CHAIN_ID not set, defaulting to 84532 (Base Sepolia)");
  }
}

validateEnvironment();

const system = `You are a helpful agent with a WaaP (Wallet as a Protocol) wallet that can interact onchain
using the Coinbase Developer Platform AgentKit. Your wallet uses two-party computation (2PC)
for key security - private keys are never fully exposed in any single location.

You are empowered to interact onchain using your tools. If you ever need funds, you can request
them from the faucet if you are on network ID 'base-sepolia'. If not, you can provide your wallet
details and request funds from the user. Before executing your first action, get the wallet details
to see what network you're on. If there is a 5XX (internal) HTTP error code, ask the user to try
again later. If someone asks you to do something you can't do with your currently available tools,
you must say so, and encourage them to implement it themselves using the CDP SDK + AgentKit,
recommend they go to docs.cdp.coinbase.com for more information. Be concise and helpful with your
responses. Refrain from restating your tools' descriptions unless it is explicitly requested.`;

const signingInstructions = `When the user asks to sign a message or sign EIP-712 typed data, you must call the appropriate signing tool directly.
Do not refuse unless a tool call actually fails. If a signing tool fails, return the exact error and ask for corrected input.`;

/**
 * Initializes AgentKit with WaaP wallet and actions.
 *
 * @returns Initialized Vercel AI SDK tools.
 */
async function initializeAgent() {
  try {
    const chainId = process.env.WAAP_CHAIN_ID || "84532";

    const walletProvider = WaapWalletProvider.configureWithWallet({
      cliPath: process.env.WAAP_CLI_PATH,
      chainId,
      rpcUrl: process.env.WAAP_RPC_URL,
      email: process.env.WAAP_EMAIL,
      password: process.env.WAAP_PASSWORD,
    });

    console.log(`WaaP wallet address: ${walletProvider.getAddress()}`);
    console.log(`Network: ${JSON.stringify(walletProvider.getNetwork())}`);

    const agentKit = await AgentKit.from({
      walletProvider,
      actionProviders: [
        wethActionProvider(),
        pythActionProvider(),
        walletActionProvider(),
        erc20ActionProvider(),
        waapAdvancedActions,
      ],
    });

    const tools = getVercelAITools(agentKit);

    const exportedWallet = walletProvider.exportWallet();
    fs.writeFileSync(WALLET_DATA_FILE, JSON.stringify(exportedWallet));

    return { tools };
  } catch (error) {
    console.error("Failed to initialize agent:", error);
    throw error;
  }
}

/**
 * Runs interactive chat mode.
 *
 * @param tools - Vercel AI SDK tools.
 */
async function runChatMode(tools: ToolSet) {
  console.log("Starting chat mode... Type 'exit' to end.");

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  const question = (prompt: string): Promise<string> =>
    new Promise(resolve => rl.question(prompt, resolve));

  const messages: Parameters<typeof streamText>[0]["messages"] = [];

  try {
    // eslint-disable-next-line no-constant-condition
    while (true) {
      const userInput = await question("\nPrompt: ");
      console.log("-------------------");

      if (userInput.toLowerCase() === "exit") {
        break;
      }

      messages.push({ role: "user", content: userInput });

      const result = streamText({
        model: openai.chat("gpt-4o-mini"),
        messages,
        tools,
        system: `${system}\n\n${signingInstructions}`,
        stopWhen: stepCountIs(10),
        onStepFinish: async ({ toolResults }) => {
          for (const tr of toolResults) {
            console.log(`Tool ${tr.toolName}: ${tr.output}`);
          }
        },
      });

      let fullResponse = "";
      for await (const delta of result.textStream) {
        fullResponse += delta;
      }

      if (fullResponse) {
        console.log("\n Response: " + fullResponse);
      }

      messages.push({ role: "assistant", content: fullResponse });

      console.log("-------------------");
    }
  } catch (error) {
    console.error("Error:", error);
  } finally {
    rl.close();
  }
}

/**
 * Runs autonomous mode on an interval.
 *
 * @param tools - Vercel AI SDK tools.
 * @param interval - Seconds between autonomous steps.
 */
async function runAutonomousMode(tools: ToolSet, interval = 10) {
  console.log("Starting autonomous mode...");

  const messages: Parameters<typeof streamText>[0]["messages"] = [];

  // eslint-disable-next-line no-constant-condition
  while (true) {
    try {
      const thought =
        "Be creative and do something interesting on the blockchain. " +
        "Choose an action or set of actions and execute it that highlights your abilities.";

      messages.push({ role: "user", content: thought });

      const result = streamText({
        model: openai.chat("gpt-4o-mini"),
        messages,
        tools,
        system: `${system}\n\n${signingInstructions}`,
        stopWhen: stepCountIs(10),
        onStepFinish: async ({ toolResults }) => {
          for (const tr of toolResults) {
            console.log(`Tool ${tr.toolName}: ${tr.output}`);
          }
        },
      });

      let fullResponse = "";
      for await (const delta of result.textStream) {
        fullResponse += delta;
      }

      if (fullResponse) {
        console.log("\n Response: " + fullResponse);
      }

      messages.push({ role: "assistant", content: fullResponse });

      console.log("-------------------");

      await new Promise(resolve => setTimeout(resolve, interval * 1000));
    } catch (error) {
      if (error instanceof Error) {
        console.error("Error:", error.message);
      }
      process.exit(1);
    }
  }
}

/**
 * Prompts user to choose chat or autonomous mode.
 *
 * @returns Selected execution mode.
 */
async function chooseMode(): Promise<"chat" | "auto"> {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  const question = (prompt: string): Promise<string> =>
    new Promise(resolve => rl.question(prompt, resolve));

  // eslint-disable-next-line no-constant-condition
  while (true) {
    console.log("\nAvailable modes:");
    console.log("1. chat    - Interactive chat mode");
    console.log("2. auto    - Autonomous action mode");

    const choice = (await question("\nChoose a mode (enter number or name): "))
      .toLowerCase()
      .trim();

    if (choice === "1" || choice === "chat") {
      rl.close();
      return "chat";
    } else if (choice === "2" || choice === "auto") {
      rl.close();
      return "auto";
    }
    console.log("Invalid choice. Please try again.");
  }
}

/**
 * Main entry point.
 */
async function main() {
  try {
    const { tools } = await initializeAgent();
    const mode = await chooseMode();

    if (mode === "chat") {
      await runChatMode(tools);
    } else {
      await runAutonomousMode(tools);
    }
  } catch (error) {
    if (error instanceof Error) {
      console.error("Error:", error.message);
    }
    process.exit(1);
  }
}

if (require.main === module) {
  console.log("Starting WaaP Agent...");
  main().catch(error => {
    console.error("Fatal error:", error);
    process.exit(1);
  });
}
